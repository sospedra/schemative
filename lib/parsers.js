'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDefault = exports.createPropTypes = exports.recursiveValues = exports.selectRecursiveStrategy = exports.recursiveArray = exports.recursiveObject = exports.getPropType = undefined;

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var getPropType = exports.getPropType = function getPropType(prop, attr) {
  return prop.iterable ? recursiveValues(prop, attr) : prop[attr];
};

var recursiveObject = exports.recursiveObject = function recursiveObject(values, attr) {
  return _lodash2.default.reduce(values, function (memo, prop, key) {
    return Object.assign(_defineProperty({}, key, getPropType(prop, attr)), memo);
  }, {});
};

var recursiveArray = exports.recursiveArray = function recursiveArray(values, attr) {
  return _lodash2.default.map(values, function (prop) {
    return getPropType(prop, attr);
  });
};

var selectRecursiveStrategy = exports.selectRecursiveStrategy = function selectRecursiveStrategy(values, attr) {
  return _lodash2.default.isArray(values) ? recursiveArray(values, attr) : recursiveObject(values, attr);
};

var recursiveValues = exports.recursiveValues = function recursiveValues(_ref, attr) {
  var type = _ref.type,
      values = _ref.values;

  var exec = attr === 'type' ? type : _lodash2.default.identity;
  return exec(selectRecursiveStrategy(values, attr));
};

var createPropTypes = exports.createPropTypes = function createPropTypes(definition) {
  return selectRecursiveStrategy(definition, 'type');
};

var createDefault = exports.createDefault = function createDefault(definition) {
  return selectRecursiveStrategy(definition, 'default');
};
//# sourceMappingURL=parsers.js.map