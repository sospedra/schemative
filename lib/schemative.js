'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createSchema = exports.shape = exports.oneOfType = exports.oneOf = exports.objectOf = exports.arrayOf = exports.symbol = exports.string = exports.object = exports.node = exports.number = exports.instanceOf = exports.func = exports.element = exports.bool = exports.array = exports.any = undefined;

var _react = require('react');

var _util = require('lodash/util');

var _util2 = _interopRequireDefault(_util);

var _parsers = require('./parsers');

var _transform = require('./transform');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var boundIsRequire = function boundIsRequire(prototype) {
  return Object.defineProperty(prototype, 'isRequired', {
    value: Object.assign({}, prototype, {
      type: prototype.type.isRequired
    })
  });
};

var createType = function createType(type, def) {
  return boundIsRequire({
    type: _react.PropTypes[type],
    default: def,
    iterable: false
  });
};

var createFunctionType = function createFunctionType(type) {
  return function (values) {
    return boundIsRequire({
      type: _react.PropTypes[type],
      values: values,
      iterable: true
    });
  };
};

var any = exports.any = createType('any', _util2.default.stubTrue());
var array = exports.array = createType('array', _util2.default.stubArray());
var bool = exports.bool = createType('bool', _util2.default.stubTrue());
var element = exports.element = createType('element', _util2.default.stubFalse());
var func = exports.func = createType('func', _util2.default.noop());
var instanceOf = exports.instanceOf = createType('instanceOf', _util2.default.stubFalse());
var number = exports.number = createType('number', -1);
var node = exports.node = createType('node', _util2.default.stubFalse());
var object = exports.object = createType('object', _util2.default.stubObject());
var string = exports.string = createType('string', _util2.default.stubString());
var symbol = exports.symbol = createType('symbol', Symbol);
var arrayOf = exports.arrayOf = createFunctionType('arrayOf');
var objectOf = exports.objectOf = createFunctionType('objectOf');
var oneOf = exports.oneOf = createFunctionType('oneOf');
var oneOfType = exports.oneOfType = createFunctionType('oneOfType');
var shape = exports.shape = createFunctionType('shape');

var createSchema = exports.createSchema = function createSchema(definition) {
  return {
    PropTypes: (0, _parsers.createPropTypes)(definition),
    Default: (0, _parsers.createDefault)(definition),
    transform: (0, _transform.createTransform)(definition)
  };
};
//# sourceMappingURL=schemative.js.map