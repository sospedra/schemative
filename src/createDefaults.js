import tsil from 'tsil'

export default (nodes) => {
  const defaultValues = tsil.modify(nodes, (node) => {
    return node.value
  })

  return tsil.deflatten(defaultValues)
}
