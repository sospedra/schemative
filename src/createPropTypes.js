import tsil from 'tsil'

export default (nodes) => {
  const propTypes = tsil.modify(nodes, (node) => {
    return node.propType
  })

  return tsil.deflatten(propTypes)
}
