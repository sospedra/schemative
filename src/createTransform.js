export default (nodes) => {
  // console.log('createTransform')
}
