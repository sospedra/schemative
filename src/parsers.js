import _ from 'lodash'

export const getPropType = (prop, attr) => {
  return prop.iterable ? recursiveValues(prop, attr) : prop[attr]
}

export const recursiveObject = (values, attr) => {
  return _.reduce(values, (memo, prop, key) => Object.assign({
    [key]: getPropType(prop, attr)
  }, memo), {})
}

export const recursiveArray = (values, attr) => {
  return _.map(values, (prop) => {
    return getPropType(prop, attr)
  })
}

export const selectRecursiveStrategy = (values, attr) => {
  return _.isArray(values)
    ? recursiveArray(values, attr)
    : recursiveObject(values, attr)
}

export const recursiveValues = ({ type, values }, attr) => {
  const exec = attr === 'type' ? type : _.identity
  return exec(selectRecursiveStrategy(values, attr))
}

export const createPropTypes = (definition) => {
  return recursiveObject(definition, 'type')
}

export const createDefault = (definition) => {
  return recursiveObject(definition, 'default')
}

export const filterByKeys = (definition, candidate) => {
  const keys = _.intersection(_.keys(definition), _.keys(candidate))
  return _.pick(candidate, keys)
}

export const transform = (definition, candidate, mutators) => Object.assign(
  filterByKeys(definition, candidate),
  _.reduce(mutators, (memo, value, key) => Object.assign({
    [key]: _.isFunction(value) ? value(candidate) : value,
    memo
  }), {})
)
