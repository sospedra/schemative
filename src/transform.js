import _ from 'lodash'

export const filterByKeys = (defsKeys, candidate) => {
  const keys = _.intersection(defsKeys, _.keys(candidate))
  return _.pick(candidate, keys)
}

export const createTransform = (definitions) => {
  const defsKeys = _.keys(definitions)
  let protoMutators = {}

  const transform = (candidate, mutators = protoMutators) => {
    return Object.assign(
      filterByKeys(defsKeys, candidate),
      _.reduce(mutators, (memo, value, key) => Object.assign({
        [key]: _.isFunction(value) ? value(candidate, definitions) : value
      }, memo), {})
    )
  }

  Object.defineProperty(transform, 'mutators', {
    get: () => protoMutators,
    set: (mutators) => {
      protoMutators = mutators
    }
  })

  return transform
}
