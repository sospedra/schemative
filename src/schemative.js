import { PropTypes } from 'react'
import _ from 'lodash/util'
import { createPropTypes, createDefault } from './parsers'
import { createTransform } from './transform'

const boundIsRequire = (prototype) => {
  return Object.defineProperty(prototype, 'isRequired', {
    value: Object.assign({}, prototype, {
      type: prototype.type.isRequired
    })
  })
}

const createType = (type, def) => boundIsRequire({
  type: PropTypes[type],
  default: def,
  iterable: false
})

const createFunctionType = (type) => (values) => boundIsRequire({
  type: PropTypes[type],
  values: values,
  iterable: true
})

export const any = createType('any', _.stubTrue())
export const array = createType('array', _.stubArray())
export const bool = createType('bool', _.stubTrue())
export const element = createType('element', _.stubFalse())
export const func = createType('func', _.noop())
export const instanceOf = createType('instanceOf', _.stubFalse())
export const number = createType('number', 0)
export const node = createType('node', _.stubFalse())
export const object = createType('object', _.stubObject())
export const string = createType('string', _.stubString())
export const symbol = createType('symbol', Symbol)
export const arrayOf = createFunctionType('arrayOf')
export const objectOf = createFunctionType('objectOf')
export const oneOf = createFunctionType('oneOf')
export const oneOfType = createFunctionType('oneOfType')
export const shape = createFunctionType('shape')

export const createSchema = (definition) => ({
  PropTypes: createPropTypes(definition),
  Default: createDefault(definition),
  transform: createTransform(definition)
})
