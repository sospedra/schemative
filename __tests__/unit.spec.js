/* global describe, it, expect */
import { PropTypes } from 'react'
import * as Schemative from '../src/schemative'
import * as parsers from '../src/parsers'
import * as mocks from './mocks'

describe('Suite types', () => {
  it('should have exist all the defined types', () => {
    mocks.types.forEach((type) => {
      expect(Schemative[type]).toMatchSnapshot()
    })

    mocks.functionTypes.forEach(([type]) => {
      expect(Schemative[type]).toMatchSnapshot()
    })
  })

  it('should be compatible with React.PropTypes', () => {
    mocks.types.forEach((type) => {
      expect(Schemative[type].propTypes).toEqual(PropTypes[type])
    })

    mocks.functionTypes.forEach(([type, baseType]) => {
      expect(Schemative[type](baseType).propTypes).toEqual(PropTypes[type])
    })
  })
})

describe('Suite createSchema', () => {
  it('should create from a simple non-nested definition', () => {
    expect(Schemative.createSchema(mocks.defs.simple)).toMatchSnapshot()
  })

  it('should create from a complex non-nested definition', () => {
    expect(Schemative.createSchema(mocks.defs.complex)).toMatchSnapshot()
  })

  it('should create from a simple nested definition', () => {
    expect(mocks.defs.simpleNested).toMatchSnapshot()
  })

  it('should create from a complex nested definition', () => {
    expect(mocks.defs.complexNested).toMatchSnapshot()
  })
})

describe('Suite parsers', () => {
  const {
    createSchemaByObject,
    createSchemaByArray,
    selectRecursiveStrategy,
    createPropTypes,
    createDefault,
    filterByKeys,
    transform
  } = parsers

  it('should create an schema given an object', () => {
    expect(createSchemaByObject(mocks.defs.object)).toMatchSnapshot()
  })

  it('should create an schema given an array', () => {
    expect(createSchemaByArray(mocks.defs.array)).toMatchSnapshot()
  })

  it('should select an strategy for recursivity', () => {
    expect(selectRecursiveStrategy(mocks.defs.array)).toMatchSnapshot()
    expect(selectRecursiveStrategy(mocks.defs.object)).toMatchSnapshot()
  })

  it('should create the schema prop types', () => {
    const propTypes = createPropTypes(mocks.defs.schemativePropTypes)
    expect(propTypes).toMatchSnapshot()
    expect(propTypes).toEqual(mocks.defs.reactPropTypes)
  })

  it('should create the schema default values', () => {
    expect(createDefault(mocks.defs.schemativePropTypes)).toMatchSnapshot()
  })

  it('should filter by keys two objects', () => {
    expect(filterByKeys(mocks.defs.simple, mocks.defs.complex)).toEqual(mocks.defs.simple)
  })

  it('should transform a definition into a fulfilled object', () => {
    expect(transform(mocks.defs.simple, mocks.fulfilled)).toEqual(mocks.fulfilled)
  })

  it('should transform a definition using mutators', () => {
    expect(transform(mocks.defs.simple, mocks.fulfilled, mocks.mutators)).toMatchSnapshot()
  })
})

describe('Suite schema', () => {
  const schema = Schemative.createSchema(mocks.defs.schemativePropTypes)

  it('should return the PropTypes', () => {
    expect(schema.PropTypes).toEqual(mocks.defs.reactPropTypes)
    expect(schema.PropTypes).toMatchSnapshot()
  })

  it('should return the default values', () => {
    expect(schema.Default).toMatchSnapshot()
  })

  it('should return a new fulfilled object after transform', () => {
    expect(schema.transform(mocks.fulfilled)).toMatchSnapshot()
  })

  it('should return a new fulfilled object with mutators', () => {
    expect(schema.transform(mocks.fulfilled, mocks.mutators)).toMatchSnapshot()
  })
})
