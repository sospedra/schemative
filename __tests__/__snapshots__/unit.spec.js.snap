exports[`Suite createSchema should create from a complex nested definition 1`] = `
Object {
  "geo": Object {
    "iterable": true,
    "propTypes": [Function],
    "value": Object {
      "coord": Object {
        "iterable": false,
        "propTypes": [Function],
        "value": -1,
      },
      "country": Object {
        "iterable": true,
        "propTypes": [Function],
        "value": Object {
          "id": Object {
            "iterable": false,
            "propTypes": [Function],
            "value": -1,
          },
        },
      },
      "flag": Object {
        "iterable": true,
        "propTypes": [Function],
        "value": Array [
          Object {
            "iterable": false,
            "propTypes": [Function],
            "value": "",
          },
          Object {
            "iterable": false,
            "propTypes": [Function],
            "value": "",
          },
        ],
      },
    },
  },
}
`;

exports[`Suite createSchema should create from a complex non-nested definition 1`] = `
Object {
  "Default": Object {
    "contacts": Array [],
    "geo": Object {},
    "getName": undefined,
    "name": "",
    "year": -1,
  },
  "PropTypes": Object {
    "contacts": [Function],
    "geo": [Function],
    "getName": [Function],
    "name": [Function],
    "year": [Function],
  },
  "transform": [Function],
}
`;

exports[`Suite createSchema should create from a simple nested definition 1`] = `
Object {
  "contact": Object {
    "iterable": true,
    "propTypes": [Function],
    "value": Array [
      Object {
        "iterable": false,
        "propTypes": [Function],
        "value": -1,
      },
      Object {
        "iterable": false,
        "propTypes": [Function],
        "value": -1,
      },
    ],
  },
  "name": Object {
    "iterable": false,
    "propTypes": [Function],
    "value": "",
  },
}
`;

exports[`Suite createSchema should create from a simple non-nested definition 1`] = `
Object {
  "Default": Object {
    "name": "",
    "year": -1,
  },
  "PropTypes": Object {
    "name": [Function],
    "year": [Function],
  },
  "transform": [Function],
}
`;

exports[`Suite parsers should create an schema given an array 1`] = `Array []`;

exports[`Suite parsers should create an schema given an object 1`] = `Object {}`;

exports[`Suite parsers should create the schema default values 1`] = `
Object {
  "arr": undefined,
  "dispatch": undefined,
  "func": undefined,
  "shape": Object {
    "deep": Object {
      "nested": "",
    },
    "id": -1,
    "object": Object {},
  },
  "value": -1,
}
`;

exports[`Suite parsers should select an strategy for recursivity 1`] = `Object {}`;

exports[`Suite parsers should select an strategy for recursivity 2`] = `Object {}`;

exports[`Suite parsers should transform a definition using mutators 1`] = `
Object {
  "memo": Object {
    "extra": "attribute",
    "memo": Object {},
  },
  "name": "Ada Lovelace",
  "year": 1852,
}
`;

exports[`Suite schema should return a new fulfilled object after transform 1`] = `Object {}`;

exports[`Suite schema should return a new fulfilled object with mutators 1`] = `
Object {
  "memo": Object {
    "extra": "attribute",
    "memo": Object {},
  },
  "year": 1852,
}
`;

exports[`Suite schema should return the default values 1`] = `
Object {
  "arr": undefined,
  "dispatch": undefined,
  "func": undefined,
  "shape": Object {
    "deep": Object {
      "nested": "",
    },
    "id": -1,
    "object": Object {},
  },
  "value": -1,
}
`;

exports[`Suite types should have exist all the defined types 1`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": Array [],
}
`;

exports[`Suite types should have exist all the defined types 2`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": true,
}
`;

exports[`Suite types should have exist all the defined types 3`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": undefined,
}
`;

exports[`Suite types should have exist all the defined types 4`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": -1,
}
`;

exports[`Suite types should have exist all the defined types 5`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": Object {},
}
`;

exports[`Suite types should have exist all the defined types 6`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": "",
}
`;

exports[`Suite types should have exist all the defined types 7`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": [Function],
}
`;

exports[`Suite types should have exist all the defined types 8`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": true,
}
`;

exports[`Suite types should have exist all the defined types 9`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": false,
}
`;

exports[`Suite types should have exist all the defined types 10`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": false,
}
`;

exports[`Suite types should have exist all the defined types 11`] = `
Object {
  "iterable": false,
  "propTypes": [Function],
  "value": false,
}
`;

exports[`Suite types should have exist all the defined types 12`] = `[Function]`;

exports[`Suite types should have exist all the defined types 13`] = `[Function]`;

exports[`Suite types should have exist all the defined types 14`] = `[Function]`;

exports[`Suite types should have exist all the defined types 15`] = `[Function]`;

exports[`Suite types should have exist all the defined types 16`] = `[Function]`;
